ca/ocsp.go

```diff
@@ -0,0 +1,125 @@
+package ca
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+// ocspLogQueue accumulates OCSP logging events and writes several of them
+// in a single log line. This reduces the number of log lines and bytes,
+// which would otherwise be quite high. As of Jan 2021 we do approximately
+// 550 rps of OCSP generation events. We can turn that into about 5.5 rps
+// of log lines if we accumulate 100 entries per line, which amounts to about
+// 3900 bytes per log line.
+// Summary of log line usage:
+// serial in hex: 36 bytes, separator characters: 2 bytes, status: 1 byte
+// If maxLogLen is 0, do not perform any accumulation or logging.
+type ocspLogQueue struct {
+	// Maximum length, in bytes, of a single log line.
+	maxLogLen int
+	// Maximum amount of time between OCSP logging events.
+	period time.Duration
+	queue  chan ocspLog
+	// This allows the stop() function to block until we've drained the queue.
+	wg     sync.WaitGroup
+	depth  prometheus.Gauge
+	logger blog.Logger
+	clk    clock.Clock
+}
+
+type ocspLog struct {
+	serial []byte
+	time   time.Time
+	status ocsp.ResponseStatus
+}
+
+func newOCSPLogQueue(
+	maxLogLen int,
+	period time.Duration,
+	stats prometheus.Registerer,
+	logger blog.Logger,
+) *ocspLogQueue {
+	depth := prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "ocsp_log_queue_depth",
+			Help: "Number of OCSP generation log entries waiting to be written",
+		})
+	stats.MustRegister(depth)
+	olq := ocspLogQueue{
+		maxLogLen: maxLogLen,
+		period:    period,
+		queue:     make(chan ocspLog, 1000),
+		wg:        sync.WaitGroup{},
+		depth:     depth,
+		logger:    logger,
+		clk:       clock.Default(),
+	}
+	olq.wg.Add(1)
+	return &olq
+}
+
+func (olq *ocspLogQueue) enqueue(serial []byte, time time.Time, status ocsp.ResponseStatus) {
+	olq.queue <- ocspLog{
+		serial: append([]byte{}, serial...),
+		time:   time,
+		status: ocsp.ResponseStatus(status),
```

Comment:
Is this cast redundant, or is it a copy-by-value thing? `ocsp.ResponseStatus` isn't a reference type so I think you're getting copy-by-value anyway even without this construction.

Reply:
This is because I chose to use the type `ocsp.ResponseStatus` for this field, but `status` is actually an int. Go requires us to be explicit about the cast.

Reply:
In the function signature here you're already declaring `status` to be type `ocsp.ResponseStatus`.

Reply:
Ah, thanks for spotting that! This code was originally in a different function that did require the cast.